docs/internal/DistributedArchitectureGuide.md

```diff
@@ -10,20 +10,58 @@
-value -- the `onResponse()` method --, and then that block of code (the `ActionListener`) is passed into a function that will eventually
-execute the code (call `onResponse()`) when a response value is available. `ActionListener`s are used to pass code down to act on a result,
-rather than lower layers returning a result back up to be acted upon by the caller. One of three things can happen to a listener: it can be
-executed in the same thread — e.g. `ActionListener.run()` --; it can be passed off to another thread to be executed; or it can be added to
-a list someplace, to eventually be executed by some service. `ActionListener`s also define `onFailure()` logic, in case an error is
-encountered before a result can be formed.
+Callbacks are used extensively throughout Elasticsearch because they enable us to write asynchronous and nonblocking code, i.e. code which
+doesn't necessarily compute a result straight away but also doesn't block the calling thread waiting for the result to become available.
+Callbacks can be passed into methods for the method to perhaps complete straight away, or instead perhaps to store in a data structure and
+complete at some later time, or (most commonly) pass along down into some other method which takes a callback argument, possibly after
+modifying its behaviour slightly by wrapping it in another callback.
+
+`ActionListener` is a general-purpose callback interface that is used extensively across the Elasticsearch codebase. It's an important
+feature of the Elasticsearch codebase that `ActionListener` is used pretty much everywhere that needs to perform some asynchronous and
+nonblocking computation. The uniformity makes it easier to compose parts of the system together without needing to build adapters to convert
+back and forth between different kinds of callback. It also makes it easier to develop the skills needed to read and understand all the
+asynchronous code, although this definitely takes practice and is certainly not easy in an absolute sense. Finally, it has allowed us to
+build a rich library for working with `ActionListener` instances themselves, creating new instances out of existing ones and completing them
+in interesting ways. See for instance:
+
+- all the static methods on [ActionListener](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java) itself
+- [`ThreadedActionListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/ThreadedActionListener.java) for forking work elsewhere
+- [`RefCountingListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/RefCountingListener.java) for running work in parallel
+- [`SubscribableListener`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java) for constructing flexible workflows
+
```

Comment:
Since there's quite a few links now throughout, you could also use a markdown feature to defer the URLs. Definitely not necessary, but I don't know if you know about it.

Put here:
[ActionListener][]

[SubscribableListener][]

Add at the end of the section (it won't be visible if you separate it by a line from other text)
[ActionListener]: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionListener.java
[SubscribableListener]: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

Reply:
Yeah, I find that makes it harder to read the plaintext (and makes no difference to the rendered copy), although neither is particularly usable. The links to code throughout are very much indicating to me that this information should be in the `ActionListener` Javadocs which has proper support for code links.